File: 0010-regular-expression-matching/0010-regular-expression-matching.js
const isMatch = (s, p) => {
    // tokenize pattern, remove duplicates
    const tokens = [];
    for (let i = 0, j = -1; i < p.length; ++i) {
        const char = p[i] === '.' ? null : p[i];
        if (char === '*') {
            if ( j > 0 && tokens[j].char === tokens[j - 1]?.char && tokens[j - 1].count === null) { // skip duplicate
                tokens.pop();
                --j;
            } else {
                tokens[j].count = null;
            }
        } else if (j > -1 && tokens[j].char === char && tokens[j].count !== null && p[i + 1] !== '*') { // skip duplicate
			++tokens[j].count;
		} else {
			tokens.push({char: char, count: 1});
			++j;
		}
    }
    
    // match string to tokens
    const match = (str, tokens, pointer, count = null) => {
        const token = tokens[pointer],
              tokenCount = count || token?.count;
        
        if (!token) {
            return !str.length; // tokens and pattern must ends at the same position
        }
        
        if (tokenCount === null) { // any match
            if (match(str, tokens, pointer + 1)) { // try skip token
                return true;
            }
            
            let i = str.length + 1;
            while (--i) {
                if (match(str, tokens, pointer, i)) { // try same token but greedy descending count
                    return true;
                }
            }
            return false;
        } else { // long match
            if (tokenCount > str.length) { // string too short
                return false;
            }
            
            if (token.char === null) { // continue testing
                return match(str.substr(tokenCount), tokens, pointer + 1);
            } else {
                let substr = str.substr(0, tokenCount),
                    pattern = token.char.padStart(tokenCount, token.char);
                
                if (substr !== pattern) { // not matched
                    return false;
                } else { // continue testing
                    return match(str.substr(tokenCount), tokens, pointer + 1);
                }   
            }
        }
    } 
    
    return match(s, tokens, 0);
};

File: 0010-regular-expression-matching/0010-regular-expression-matching.go
type Token struct {
  Rune rune
  Count int
}

func getPatternTokens(p string) *[]Token {
  tokens := []Token{}
  p += " "
  for i, rn := range p[:len(p) - 1] {
    ln := len(tokens)
    var tail, pretail *Token
    if ln > 0 { tail = &tokens[ln - 1] }
    if ln > 1 { pretail = &tokens[ln - 2] }

    if (rn == '*') {
      if (pretail != nil && tail.Rune == pretail.Rune && pretail.Count == 0) { // skip duplicate
        tokens = tokens[:ln - 1]
      } else {
        tail.Count = 0
      }
    } else if (tail != nil && tail.Rune == rn && tail.Count > 0 && p[i + 1] != '*') { // skip duplicate
      tail.Count++
    } else {
      tokens = append(tokens, Token{ rn, 1 })
    }
  }
  return &tokens
}

func matchTokens(s string, tokens *[]Token, pointer int, tokenCount int) bool {
  if pointer >= len(*tokens) {
    // tokens and pattern must ends at the same position
    return len(s) == 0
  }

  token := &(*tokens)[pointer]
  if tokenCount == 0 { tokenCount = token.Count }
  
  if tokenCount == 0 { // any match
    if matchTokens(s, tokens, pointer + 1, 0) { // try skip token
      return true
    }  
    for i := len(s); i > 0; i-- {
      if matchTokens(s, tokens, pointer, i) { // try same token but greedy descending count
        return true
      }
    }
    return false
  } else { // long match
    if tokenCount > len(s) { // string too short
      return false
    }
    if token.Rune == '.' { // continue testing
      return matchTokens(s[tokenCount:], tokens, pointer + 1, 0);
    } else {
      substr := s[:tokenCount]
      pattern := strings.Repeat(string(token.Rune), tokenCount)
      
      if substr != pattern { // not matched
        return false
      } else { // continue testing
        return matchTokens(s[tokenCount:], tokens, pointer + 1, 0);
      }   
    }
  }
}

func isMatch(s string, p string) bool {
  tokens := getPatternTokens(p)
  return matchTokens(s, tokens, 0, 0)
}